File: apps/node/src/chapter09/main.ts
import { join } from "node:path";
import { dataDir } from "../dataset.ts";
import {
  joinWeather,
  loadBike,
  loadWeather,
  meanCountByWeather,
} from "./bike-weather.ts";
import { prepareBoston, scoreFeatureSet } from "./boston.ts";
import { iqrOutliers, removeTargetOutliers } from "./outliers.ts";
import { pairsWithReplacement, termName } from "./polynomial-features.ts";
import { Standardizer } from "./standardizer.ts";

const TEST_SIZE = 0.3;
const SEED = 0;
// 浮動小数点の誤差で -0.00 と表示されないように、ごく小さい値は 0 とみなす
const ZERO_TOLERANCE = 1e-9;
const SCORE_DIGITS = 4;
const MEAN_DIGITS = 2;
const COUNT_DIGITS = 1;

export const COLUMNS = ["RM", "LSTAT", "PTRATIO"];
export const SQUARES = ["RM^2", "LSTAT^2", "PTRATIO^2"];
export const FEATURE_SETS = new Map([
  ["元の特徴量", COLUMNS],
  ["2 乗の項を追加", [...COLUMNS, ...SQUARES]],
  [
    "交互作用の項も追加",
    [
      ...COLUMNS,
      ...pairsWithReplacement(COLUMNS).map(([left, right]) =>
        termName(left, right),
      ),
    ],
  ],
]);

function format(value: number, digits: number): string {
  return (Math.abs(value) < ZERO_TOLERANCE ? 0 : value).toFixed(digits);
}

function formatScores([train, test]: [number, number]): string {
  return `訓練 ${format(train, SCORE_DIGITS)}, テスト ${format(test, SCORE_DIGITS)}`;
}

export function main(print: (line: string) => void = console.log): void {
  const split = prepareBoston(join(dataDir(), "Boston.csv"), TEST_SIZE, SEED);
  print(
    `訓練データ: ${split.xTrain.length} 件, テストデータ: ${split.xTest.length} 件`,
  );
  print(`特徴量の列: ${Object.keys(split.xTrain[0] ?? {}).join(", ")}`);

  // 標準化した後の平均と標準偏差も、Standardizer.fit で求める
  const standardized = Standardizer.fit(split.xTrain, ["RM"]).transform(
    split.xTrain,
  );
  const rm = Standardizer.fit(standardized, ["RM"]);
  print(
    `標準化した訓練データの RM: 平均 ${format(rm.means.RM, MEAN_DIGITS)}, 標準偏差 ${format(rm.stds.RM, MEAN_DIGITS)}`,
  );

  print("決定係数:");
  for (const [name, terms] of FEATURE_SETS) {
    print(
      `  ${name}（${terms.length} 列）: ${formatScores(scoreFeatureSet(split, COLUMNS, terms))}`,
    );
  }

  const outliers = iqrOutliers(split.tTrain).filter((outlier) => outlier);
  print(`訓練データの PRICE の外れ値: ${outliers.length} 件`);
  const removed = scoreFeatureSet(removeTargetOutliers(split), COLUMNS, [
    ...COLUMNS,
    ...SQUARES,
  ]);
  print(`  外れ値を除いて 2 乗の項を追加: ${formatScores(removed)}`);

  const joined = joinWeather(
    loadBike(join(dataDir(), "bike.tsv")),
    loadWeather(join(dataDir(), "weather.csv")),
  );
  const means = [...meanCountByWeather(joined)]
    .map(([weather, count]) => `${weather}=${format(count, COUNT_DIGITS)}`)
    .join(", ");
  print(`天気ごとの平均利用者数: ${means}`);
}

// node src/chapter09/main.ts で直接実行したときだけ main を呼ぶ
if (import.meta.filename === process.argv[1]) {
  main();
}

File: apps/java/src/main/java/chapter09/Main.java
package chapter09;

import chapter02.Features;
import chapter02.TrainTestSplit;
import dataset.DataDir;
import java.io.IOException;
import java.nio.file.Path;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/** ボストンの住宅価格で特徴量エンジニアリングを試し、自転車の利用者数に天気を結合して集計する。 */
public final class Main {
  /** 多項式特徴量を作る元の列 */
  public static final List<String> COLUMNS = List.of("RM", "LSTAT", "PTRATIO");

  /** 2 乗の項 */
  public static final List<String> SQUARES = List.of("RM^2", "LSTAT^2", "PTRATIO^2");

  private static final double TEST_SIZE = 0.3;
  private static final long SEED = 0;

  // 浮動小数点の誤差で -0.00 と表示されないように、ごく小さい値は 0 とみなす
  private static final double ZERO_TOLERANCE = 1e-9;

  private static final int SCORE_DIGITS = 4;
  private static final int MEAN_DIGITS = 2;
  private static final int COUNT_DIGITS = 1;

  private Main() {}

  /** 特徴量の組の名前と、使う項。 */
  public static Map<String, List<String>> featureSets() {
    List<String> interactions =
        PolynomialFeatures.pairsWithReplacement(COLUMNS).stream()
            .map(PolynomialFeatures.Pair::name)
            .toList();
    Map<String, List<String>> sets = new LinkedHashMap<>();
    sets.put("元の特徴量", COLUMNS);
    sets.put("2 乗の項を追加", concat(COLUMNS, SQUARES));
    sets.put("交互作用の項も追加", concat(COLUMNS, interactions));
    return sets;
  }

  private static List<String> concat(List<String> first, List<String> second) {
    return Stream.concat(first.stream(), second.stream()).toList();
  }

  public static void main(String[] args) throws IOException {
    Path dataDir = DataDir.dataDir();
    TrainTestSplit<Features, Double> split =
        Boston.prepare(dataDir.resolve("Boston.csv"), TEST_SIZE, SEED);
    System.out.println(
        "訓練データ: " + split.xTrain().size() + " 件, テストデータ: " + split.xTest().size() + " 件");
    System.out.println("特徴量の列: " + String.join(", ", split.xTrain().getFirst().columns()));

    // 標準化した訓練データの平均と標準偏差を、もう一度 fit して確かめる
    Standardizer check =
        Standardizer.fit(Standardizer.fit(split.xTrain()).transform(split.xTrain()));
    System.out.println(
        "標準化した訓練データの RM: 平均 "
            + format(check.means().get("RM"), MEAN_DIGITS)
            + ", 標準偏差 "
            + format(check.stds().get("RM"), MEAN_DIGITS));

    System.out.println("決定係数:");
    featureSets()
        .forEach(
            (name, terms) ->
                System.out.println(
                    "  "
                        + name
                        + "（"
                        + terms.size()
                        + " 列）: "
                        + formatScores(Boston.scoreFeatureSet(split, COLUMNS, terms))));

    long outliers = Outliers.iqrOutliers(split.tTrain()).stream().filter(b -> b).count();
    System.out.println("訓練データの PRICE の外れ値: " + outliers + " 件");
    Scores removed =
        Boston.scoreFeatureSet(
            Outliers.removeTargetOutliers(split), COLUMNS, concat(COLUMNS, SQUARES));
    System.out.println("  外れ値を除いて 2 乗の項を追加: " + formatScores(removed));

    var joined =
        BikeWeather.joinWeather(
            BikeWeather.loadBike(dataDir.resolve("bike.tsv")),
            BikeWeather.loadWeather(dataDir.resolve("weather.csv")));
    System.out.println(
        "天気ごとの平均利用者数: "
            + BikeWeather.meanCountByWeather(joined).entrySet().stream()
                .map(e -> e.getKey() + "=" + format(e.getValue(), COUNT_DIGITS))
                .collect(Collectors.joining(", ")));
  }

  private static String format(double value, int digits) {
    return String.format(
        Locale.ROOT, "%." + digits + "f", Math.abs(value) < ZERO_TOLERANCE ? 0.0 : value);
  }

  private static String formatScores(Scores scores) {
    return "訓練 "
        + format(scores.train(), SCORE_DIGITS)
        + ", テスト "
        + format(scores.test(), SCORE_DIGITS);
  }
}
